zng decoder crash on bad data

The zng decoder crashes ungracefully with a certain bad input created as a zst file.  It should detect the problem and report an error.

Repro is at Zed commit 520b145 with this `t.zson`...
```
{a:[1,2],b:[]([int64])}
{a:null([int64]),b:[1,2,3]}
```
Make a zst file from it then run zq on the zst as if it were zng either via autodetection or with `-i zng`...
```
$ zq -version
Version: v1.2.0-48-g520b145a

$ zq -f zst -o t.zst t.zson
$ zq -i zng t.zst

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1150c15]

goroutine 60 [running]:
github.com/brimdata/zed.appendTypeValue({0xc0003dc740, 0x1, 0x40}, {0x0?, 0x0?}, 0x0?)
	/Users/mccanne/repo/zed/type.go:536 +0x515
github.com/brimdata/zed.appendTypeValue({0xc0003dc740, 0x0, 0x40}, {0x1921c50?, 0xc000392990?}, 0xc000404c88)
	/Users/mccanne/repo/zed/type.go:515 +0x227
...
```

--- a/context.go
+++ b/context.go
@@ -63,7 +63,7 @@ func (c *Context) LookupType(id int) (Type, error) {
 		return nil, fmt.Errorf("type id (%d) cannot be negative", id)
 	}
 	if id < IDTypeComplex {
-		return LookupPrimitiveByID(id), nil
+		return LookupPrimitiveByID(id)
 	}
 	c.mu.RLock()
 	defer c.mu.RUnlock()
@@ -514,12 +514,8 @@ func (c *Context) DecodeTypeValue(tv zcode.Bytes) (Type, zcode.Bytes) {
 		}
 		return typ, tv
 	default:
-		if id < 0 || id > TypeValueMax {
-			// Out of range.
-			return nil, nil
-		}
-		typ := LookupPrimitiveByID(int(id))
-		if typ == nil {
+		typ, err := LookupPrimitiveByID(int(id))
+		if err != nil {
 			return nil, nil
 		}
 		return typ, tv

--- a/mapper.go
+++ b/mapper.go
@@ -24,7 +24,8 @@ func NewMapper(out *Context) *Mapper {
 // does not change anything.
 func (m *Mapper) Lookup(id int) Type {
 	if id < IDTypeComplex {
-		return LookupPrimitiveByID(id)
+		typ, _ := LookupPrimitiveByID(id)
+		return typ
 	}
 	m.mu.RLock()
 	defer m.mu.RUnlock()
@@ -36,7 +37,7 @@ func (m *Mapper) Lookup(id int) Type {
 
 func (m *Mapper) Enter(id int, ext Type) (Type, error) {
 	if id < IDTypeComplex {
-		return LookupPrimitiveByID(id), nil
+		return LookupPrimitiveByID(id)
 	}
 	typ, err := m.outputCtx.TranslateType(ext)
 	if err != nil {

--- a/runtime/expr/eval.go
+++ b/runtime/expr/eval.go
@@ -423,7 +423,10 @@ func (a *Add) Eval(ectx Context, this *zed.Value) *zed.Value {
 	if zerr != nil {
 		return zerr
 	}
-	typ := zed.LookupPrimitiveByID(id)
+	typ, err := zed.LookupPrimitiveByID(id)
+	if err != nil {
+		return a.zctx.NewError(err)
+	}
 	switch {
 	case zed.IsFloat(id):
 		v1, v2 := a.operands.floats()
@@ -450,7 +453,10 @@ func (s *Subtract) Eval(ectx Context, this *zed.Value) *zed.Value {
 	if zerr != nil {
 		return zerr
 	}
-	typ := zed.LookupPrimitiveByID(id)
+	typ, err := zed.LookupPrimitiveByID(id)
+	if err != nil {
+		return s.zctx.NewError(err)
+	}
 	switch {
 	case zed.IsFloat(id):
 		v1, v2 := s.operands.floats()
@@ -473,7 +479,10 @@ func (m *Multiply) Eval(ectx Context, this *zed.Value) *zed.Value {
 	if zerr != nil {
 		return zerr
 	}
-	typ := zed.LookupPrimitiveByID(id)
+	typ, err := zed.LookupPrimitiveByID(id)
+	if err != nil {
+		return m.zctx.NewError(err)
+	}
 	switch {
 	case zed.IsFloat(id):
 		v1, v2 := m.operands.floats()
@@ -496,7 +505,10 @@ func (d *Divide) Eval(ectx Context, this *zed.Value) *zed.Value {
 	if zerr != nil {
 		return zerr
 	}
-	typ := zed.LookupPrimitiveByID(id)
+	typ, err := zed.LookupPrimitiveByID(id)
+	if err != nil {
+		return d.zctx.NewError(err)
+	}
 	switch {
 	case zed.IsFloat(id):
 		v1, v2 := d.operands.floats()
@@ -528,7 +540,10 @@ func (m *Modulo) Eval(ectx Context, this *zed.Value) *zed.Value {
 	if zerr != nil {
 		return zerr
 	}
-	typ := zed.LookupPrimitiveByID(id)
+	typ, err := zed.LookupPrimitiveByID(id)
+	if err != nil {
+		return m.zctx.NewError(err)
+	}
 	if zed.IsFloat(id) || !zed.IsNumber(id) {
 		return ectx.CopyValue(m.zctx.NewErrorf("type %s incompatible with '%%' operator", zson.FormatType(typ)))
 	}

--- a/runtime/expr/sort.go
+++ b/runtime/expr/sort.go
@@ -201,8 +201,10 @@ func compareValues(a, b *zed.Value, comparefns map[zed.Type]comparefn, pair *coe
 	abytes, bbytes := a.Bytes, b.Bytes
 	if a.Type.ID() != b.Type.ID() {
 		id, err := pair.Coerce(a, b)
-		typ = zed.LookupPrimitiveByID(id)
-		if err != nil || typ == nil {
+		if err == nil {
+			typ, err = zed.LookupPrimitiveByID(id)
+		}
+		if err != nil {
 			// If values cannot be coerced, just compare the native
 			// representation of the type.
 			// XXX This is heavyweight and should probably just compare

--- a/type.go
+++ b/type.go
@@ -261,48 +261,54 @@ func PrimitiveName(typ Type) string {
 	}
 }
 
-func LookupPrimitiveByID(id int) Type {
+func LookupPrimitiveByID(id int) (Type, error) {
+	if id < 0 {
+		return nil, fmt.Errorf("negative type ID: %d", id)
+	}
+	if id >= IDTypeComplex {
+		return nil, fmt.Errorf("type ID too large for primitive: %d", id)
+	}
 	switch id {
 	case IDBool:
-		return TypeBool
+		return TypeBool, nil
 	case IDInt8:
-		return TypeInt8
+		return TypeInt8, nil
 	case IDUint8:
-		return TypeUint8
+		return TypeUint8, nil
 	case IDInt16:
-		return TypeInt16
+		return TypeInt16, nil
 	case IDUint16:
-		return TypeUint16
+		return TypeUint16, nil
 	case IDInt32:
-		return TypeInt32
+		return TypeInt32, nil
 	case IDUint32:
-		return TypeUint32
+		return TypeUint32, nil
 	case IDInt64:
-		return TypeInt64
+		return TypeInt64, nil
 	case IDUint64:
-		return TypeUint64
+		return TypeUint64, nil
 	case IDFloat32:
-		return TypeFloat32
+		return TypeFloat32, nil
 	case IDFloat64:
-		return TypeFloat64
+		return TypeFloat64, nil
 	case IDBytes:
-		return TypeBytes
+		return TypeBytes, nil
 	case IDString:
-		return TypeString
+		return TypeString, nil
 	case IDIP:
-		return TypeIP
+		return TypeIP, nil
 	case IDNet:
-		return TypeNet
+		return TypeNet, nil
 	case IDTime:
-		return TypeTime
+		return TypeTime, nil
 	case IDDuration:
-		return TypeDuration
+		return TypeDuration, nil
 	case IDType:
-		return TypeType
+		return TypeType, nil
 	case IDNull:
-		return TypeNull
+		return TypeNull, nil
 	}
-	return nil
+	return nil, fmt.Errorf("primitive type ID %d not implemented", id)
 }
 
 // Utilities shared by complex types (ie, set and array)

--- a/zio/zngio/ztests/issue-4082.yaml
+++ b/zio/zngio/ztests/issue-4082.yaml
@@ -1,14 +1,18 @@
 script: |
-  zq -f zst -o f.zst -
-  ! zq -i zng f.zst
+  ! zq -i zng -
 
 inputs:
   - name: stdin
-    data: |
-      {a:[1,2],b:[]([int64])}
-      {a:null([int64]),b:[1,2,3]}
+    # Generated with `bash -c 'zq -f zst -o f.zst - <<<"[1,2]" && base64 -b 70 f.zst'`.
+    data: !!binary |
+      AgQCAgIEAQUFAAIGT2Zmc2V0CQZMZW5ndGgIBwdTZWdtZW50HgEfAAIEVHlwZRwGU2VnbW
+      FwIAcJUHJpbWl0aXZlIQACB0xlbmd0aHMgBlZhbHVlcyIHBUFycmF5IxcBIAYFAgwCAiQP
+      BQQBAgQJAgkGBQIEAgj/SAYAaPQKAQkAAgtza2V3X3RocmVzaAkOc2VnbWVudBAA8DgHDH
+      pzdC5GaWxlTWV0YR8ABQVtYWdpYxkEdHlwZRkHdmVyc2lvbgkIc2VjdGlvbnMeBG1ldGEg
+      Bw16bmdpby5UcmFpbGVyIRMCIiIMWk5HIFRyYWlsZXIEenN0AgYFAg4C4goFAAAgAwQAAK
+      D/
 
 outputs:
   - name: stderr
     data: |
-      f.zst: primitive type ID 4 not implemented
+      stdio:stdin: primitive type ID 4 not implemented

--- a/zio/zngio/ztests/issue-4082.yaml
+++ b/zio/zngio/ztests/issue-4082.yaml
@@ -1,0 +1,14 @@
+script: |
+  zq -f zst -o f.zst -
+  ! zq -i zng f.zst
+
+inputs:
+  - name: stdin
+    data: |
+      {a:[1,2],b:[]([int64])}
+      {a:null([int64]),b:[1,2,3]}
+
+outputs:
+  - name: stderr
+    data: |
+      f.zst: primitive type ID 4 not implemented

--- a/zson/formatter.go
+++ b/zson/formatter.go
@@ -232,11 +232,12 @@ func (f *Formatter) formatTypeValue(indent int, tv zcode.Bytes) zcode.Bytes {
 	}
 	switch n {
 	default:
-		if n < 0 || n > zed.TypeValueMax {
-			f.buildf("<ERR bad type ID %d in type value>", n)
+		typ, err := zed.LookupPrimitiveByID(int(n))
+		if err != nil {
+			f.buildf("<ERR bad type ID in type value: %s>", err)
 			return nil
 		}
-		typ := zed.LookupPrimitiveByID(int(n))
+
 		f.startColor(color.Gray(160))
 		f.build(zed.PrimitiveName(typ))
 		f.endColor()

